File: libs/ts/test/clock.test.ts
import './setup.js';
import { describe, it, afterEach } from 'node:test';
import assert from 'node:assert/strict';
import { BellSchedule } from '../src/bell-schedule.js';
import { setDebugTime, setDebugOffset, clearDebugTime, getDebugOffset } from '../src/clock.js';

const CALENDAR_DATA = {
  year: '2025-2026',
  timezone: 'America/Los_Angeles',
  firstDay: '2025-08-13',
  lastDay: '2026-06-04',
  schedules: {
    NORMAL: [
      { name: 'Period 1', start: '8:30', end: '9:28' },
      { name: 'Period 2', start: '9:34', end: '10:37' },
    ],
  },
  weekdaySchedules: {},
  holidays: [],
  teacherWorkDays: [],
  breakNames: {},
};

const bells = () => new BellSchedule([CALENDAR_DATA], { role: 'student', includeTags: {} });

// A moment inside Period 1 on Tuesday 2025-08-19 (a normal school day).
const duringPeriod1 = Temporal.ZonedDateTime.from({
  year: 2025,
  month: 8,
  day: 19,
  hour: 8,
  minute: 45,
  timeZone: 'America/Los_Angeles',
}).toInstant();

describe('debug clock', () => {
  afterEach(clearDebugTime);

  it('defaults to the real clock (no offset)', () => {
    assert.equal(getDebugOffset(), null);
  });

  it('setDebugTime makes time-defaulting methods resolve as if now were that instant', () => {
    assert.equal(bells().currentInterval(duringPeriod1)?.name, 'Period 1'); // sanity: explicit arg
    setDebugTime(duringPeriod1);
    assert.equal(bells().currentInterval()?.name, 'Period 1'); // no arg → uses debug time
    assert.equal(bells().periodAt()?.name, 'Period 1');
  });

  it('clearDebugTime restores the real clock', () => {
    setDebugTime(duringPeriod1);
    assert.notEqual(getDebugOffset(), null);
    clearDebugTime();
    assert.equal(getDebugOffset(), null);
  });

  it('setDebugOffset and setDebugTime agree', () => {
    setDebugTime(duringPeriod1);
    const viaTime = bells().currentInterval()?.name;
    const offset = getDebugOffset();
    clearDebugTime();
    assert.ok(offset);
    setDebugOffset(offset);
    assert.equal(bells().currentInterval()?.name, viaTime);
  });

  it('an explicitly-passed instant overrides the debug offset', () => {
    // Pretend it is a summer day with no school...
    const summer = Temporal.ZonedDateTime.from({
      year: 2025,
      month: 7,
      day: 15,
      hour: 12,
      timeZone: 'America/Los_Angeles',
    }).toInstant();
    setDebugTime(summer);
    assert.equal(bells().currentInterval(), null); // debug time → summer, no period
    // ...but an explicit instant still wins.
    assert.equal(bells().currentInterval(duringPeriod1)?.name, 'Period 1');
  });
});

File: libs/java/src/test/java/com/gigamonkeys/bells/ClockTest.java
package com.gigamonkeys.bells;

import static com.gigamonkeys.bells.Fixtures.SIMPLE_DATA;
import static com.gigamonkeys.bells.Fixtures.laInstant;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

import java.time.Duration;
import java.time.Instant;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

/** The process-global debug clock (DateTimes.setDebugTime / clearDebugTime / ...). */
class ClockTest {

  // A moment inside Period 1 on Tuesday 2025-08-19 (a normal school day).
  private static final Instant DURING_PERIOD_1 = laInstant("2025-08-19T08:45");

  private static BellSchedule make() {
    return new BellSchedule(CalendarData.parse(SIMPLE_DATA), Options.defaults());
  }

  @AfterEach
  void resetClock() {
    DateTimes.clearDebugTime();
  }

  @Test
  void defaultsToRealClock() {
    assertNull(DateTimes.getDebugOffset());
  }

  @Test
  void setDebugTimeDrivesTimeDefaultingMethods() {
    // Sanity: an explicit instant resolves to Period 1.
    assertEquals("Period 1", make().currentInterval(DURING_PERIOD_1).name());
    DateTimes.setDebugTime(DURING_PERIOD_1);
    // No argument -> uses the debug time.
    assertEquals("Period 1", make().currentInterval().name());
    assertEquals("Period 1", make().periodAt().name());
  }

  @Test
  void clearDebugTimeRestoresRealClock() {
    DateTimes.setDebugTime(DURING_PERIOD_1);
    assertNotNull(DateTimes.getDebugOffset());
    DateTimes.clearDebugTime();
    assertNull(DateTimes.getDebugOffset());
  }

  @Test
  void setDebugOffsetAndSetDebugTimeAgree() {
    DateTimes.setDebugTime(DURING_PERIOD_1);
    String viaTime = make().currentInterval().name();
    Duration offset = DateTimes.getDebugOffset();
    DateTimes.clearDebugTime();
    assertNotNull(offset);
    DateTimes.setDebugOffset(offset);
    assertEquals(viaTime, make().currentInterval().name());
  }

  @Test
  void explicitInstantOverridesDebugOffset() {
    // Pretend it is a summer day with no school...
    DateTimes.setDebugTime(laInstant("2025-07-15T12:00"));
    assertNull(make().currentInterval());
    // ...but an explicit instant still wins.
    assertEquals("Period 1", make().currentInterval(DURING_PERIOD_1).name());
  }

  @Test
  void setDebugOffsetAcceptsDuration() {
    DateTimes.setDebugOffset(Duration.ofHours(-3));
    assertEquals(Duration.ofHours(-3), DateTimes.getDebugOffset());
  }
}
